index.js

```diff
@@ -264,250 +212,256 @@
   $.TypeClass = function(name, test) {
     return {
       '@@type': 'sanctuary-def/TypeClass',
-      name: name,
       _test: test,
+      name: name,
       toString: always(stripNamespace(name))
     };
   };
 
-  //  testFrom :: (a -> Result) -> a -> Boolean
-  var testFrom = function(validate) {
-    return function(x) {
-      return validate(x).isRight;
+  //  createType :: ... -> Type
+  var createType = function(
+    typeName,   // :: String
+    name,       // :: String
+    format,     // :: (String -> String, String -> String -> String) -> String
+    test,       // :: a -> Boolean
+    keys,       // :: Array String
+    types       // :: StrMap { extractor :: a -> Array b, type :: Type }
+  ) {
+    var validate = function(x) {
+      if (!test(x)) return Left({value: x, propPath: []});
+      for (var idx = 0; idx < keys.length; idx += 1) {
+        var k = keys[idx];
+        var t = types[k];
+        for (var idx2 = 0, ys = t.extractor(x); idx2 < ys.length; idx2 += 1) {
+          var result = t.type.validate(ys[idx2]);
+          if (result.isLeft) {
+            var value = result.value.value;
+            var propPath = [k].concat(result.value.propPath);
```

Comment:
You may be thinking of #53. I *think* it would be okay to use local mutation here, but it would take significant mental effort to convince myself that no one else is holding a reference to the array. I don't see a significant downside to creating a new array here, particularly since we're about to short-circuit.